node/src/components/consensus/highway_core/state.rs

```diff
@@ -705,6 +695,120 @@ impl<C: Context> State<C> {
     pub(crate) fn is_empty(&self) -> bool {
         self.units.is_empty()
     }
+
+    /// Validates whether `wunit` violates the Limited Naïveté Criterion (LNC).
+    /// Returns index of the first equivocator that was cited naively in violation of the LNC, or
+    /// `None` if the LNC is satisfied.
+    fn validate_lnc(&self, wunit: &WireUnit<C>) -> Option<ValidatorIndex> {
+        wunit
+            .panorama
+            .enumerate()
+            .filter(|(_, obs)| obs.is_faulty())
+            .map(|(i, _)| i)
+            .find(|eq_idx| !self.satisfies_lnc_for(wunit, *eq_idx))
+    }
+
+    /// Returns `true` if there is at most one fork by the validator `eq_idx` that is cited naively
+    /// by `wunit` or earlier units by the same creator.
+    fn satisfies_lnc_for(&self, wunit: &WireUnit<C>, eq_idx: ValidatorIndex) -> bool {
+        // Find all forks by eq_idx that are cited naively by wunit.
+        // * If it's more than one, return false (the LNC is violated).
+        // * If it's none, return true: If the LNC were violated we would have already noticed when
+        //   validating earlier units by wunit.creator.
+        // * Otherwise store the unique naively cited fork in naive_by_wunit.
+        let mut opt_naive_by_wunit = None;
+        {
+            let seen_by_endorsed =
+                |hash| wunit.endorsed.iter().any(|e_hash| self.sees(e_hash, hash));
```

Comment:
Should we verify that `wunit.endorsed` are seen from `wunit.panorama`?

Reply:
Good point! :scream: